### PR TITLE
React 18

### PR DESCRIPTION
**Pull request by jcfilben**
#### What does this PR do?
Upgrades to React v18

- Anyone consuming Grommet should be able to use React v16, v17, or v18. The breaking changes in react 18 are within our testing files.

Important changes in React 18:
**ReactDOM.render is no longer supported**
You are still able to use ReactDOM.render but you will get a warning that it is deprecated. `createRoot` should be used instead of render. Render was being used in the following places in grommet:
- In the unit tests we are using render from `@testing-library/react`. As of v13.0.0 of `@testing-library/react` when you use render from testing-library it is actually using createRoot under the hood.
- Storybook is loading stories with render. There is a pre-release branch of storybook that supports `createRoot` but it looks like they still need to iron out some issues so we are sticking with the official release version of storybook for now. This means that when you run storybook everything should work as before but there will be warnings in the console on each story. **We will need to make a note of this in the release notes if it is not fixed before then**.
- RoutedAnchor and RoutedButton use `render`. This means if these components are used in a React 18 application they will generate a warning but still function. Since these components are deprecated I felt this was fine.

**Updates to TypeScript definitions**
The most notable change is that the `children` prop now needs to be explicitly defined. A list of the typescript changes can be seen here: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/56210. I have updated the relevant type definitions to include children. It is possible that I may have missed updating some of the types so something we will want to keep an eye on as we move closer to the release this month.

Other Notes:
- I removed the `userAgent` test from Grommet-test.js. With the testing library update this test's document.body's client width was always returning 0 so the ResponsiveContext would always receive a size of small. I ensured that the `userAgent` code in Grommet was working as expected and it was. I played around with the test but in the end felt that there wasn't a good way to write the test that was effectively testing the `userAgent`. I think if we want to have this kind of test it is better suited to an end-to-end type of testing environment instead of testing with jsdom (which acknowledges that it shouldn't be used for this kind of testing).

Helpful links:
How to upgrade to react 18 - https://reactjs.org/blog/2022/03/08/react-18-upgrade-guide.html
React's changelog - https://github.com/facebook/react/blob/main/CHANGELOG.md

#### What testing has been done on this PR?
`yarn test`
`yarn storybook`
`yarn build`
#### Do Jest tests follow these best practices?

- [ ] `screen` is used for querying.
- [ ] The correct query is used. (Refer to [this list of queries](https://testing-library.com/docs/queries/about/#priority))
- [ ] `userEvent` is used in place of `fireEvent`.
- [ ] `asFragment()` is used for snapshot testing.

#### What are the relevant issues?
Closes https://github.com/grommet/grommet/issues/5947
Closes https://github.com/grommet/grommet/issues/6062

#### Do the grommet docs need to be updated?
Yes
#### Should this PR be mentioned in the release notes?
Yes
#### Is this change backwards compatible or is it a breaking change?
Backwards compatible

### CODE REVIEW

File: src/js/components/Footer/index.d.ts [hunk 1]
```diff
@@ -1,6 +1,6 @@
 import * as React from 'react';
-import { BoxProps } from '../Box/index';
```
Comment (britt6612) [inline src/js/components/Footer/index.d.ts:2]: I believe we have the same in `Header` in the `index.d.ts` can we update that ?

File: src/js/components/Grommet/__tests__/Grommet-test.js [hunk 2]
```diff
@@ -15,35 +14,6 @@ const TestAnnouncer = ({ announce }) => {
...
-        value: 1000,
-      },
-    },
-  },
-};
-
-const SSRTester = ({ ua }) => (
-  <Grommet theme={customBreakpointsTheme} userAgent={ua}>
-    <ResponsiveContext.Consumer>
-      {(size) => <Heading>{`Received size ${size} for ${ua}`}</Heading>}
```
Comment (britt6612) [inline src/js/components/Grommet/__tests__/Grommet-test.js:42]: should we add this issue of testing to our end-to-end issue we have on the DS backlog? Just so we know when we start that we can account for this.

File: src/js/components/NameValueList/index.d.ts [hunk 3]
```diff
@@ -3,6 +3,7 @@ import { AlignType, WidthType } from '../../utils';
 import { GridProps } from '../Grid';
 export interface NameValueListProps {
   align?: AlignType;
+  children?: React.ReactNode;
```
Comment (britt6612) [inline src/js/components/NameValueList/index.d.ts:6]: more of a question in some of the files you added the ` children?: React.ReactNode;`
as well as adding `children` to the omit` Omit<ButtonType, 'icon' | 'size' | 'children'> {}`  just trying to understand why this is included in certain files?

Comment (jcfilben) [reply src/js/components/NameValueList/index.d.ts:6]: I am omitting the `children` prop and then redefining it in certain cases if the `children` prop that was given through the intrinsic element or interface has a different type definition from what we need.

File: src/js/components/Grommet/__tests__/Grommet-test.js (hunk 2 above)
Comment (jcfilben) [reply src/js/components/Grommet/__tests__/Grommet-test.js:42]: Yeah I will add this to the end-to-end testing issue in our backlog

File: src/js/components/Footer/index.d.ts (hunk 1 above)
Comment (jcfilben) [reply src/js/components/Footer/index.d.ts:2]: Yeah, Header is using `BoxTypes` which includes the `children` prop but to keep Header consistent with the other files I updated it to use `BoxExtendedProps`

File: src/js/components/NameValueList/index.d.ts (hunk 3 above)
Comment (britt6612) [reply src/js/components/NameValueList/index.d.ts:6]: thank you for clarifying

File: src/js/components/Footer/index.d.ts (hunk 1 above)
Comment (britt6612) [reply src/js/components/Footer/index.d.ts:2]: yeah that's what I meant to add in the `BoxExtendedProps` for Header thank you 😄

**Review by amandadupell — APPROVED** [review]
changes make sense based on looking into react 18 upgrades and explanations, tests look good!

**Review by ericsoderberghp — APPROVED** [review]

**Pull request merged**